keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -12,124 +11,477 @@ approvers:
+
+### Summary of the proposed changes
+
+1. Add path as a prefix and make regex support optional. The current spec states
+   that the path is a regular expression and arbitrary regular expression, but
+   support varies across providers. In addition, regex matching is not supported
+   by many of the cloud provider implementations.
+1. Fix API field naming:
+   1. `spec.backend` should be called `spec.defaultBackend`.
+1. Hostname wildcard matching. We currently allow for creation of
+   `*.foo.com` and this seems to be a commonly suppported host match,
+   but this is not part of the spec.
+1. Specify healthcheck behavior and be able to configure the healthcheck path
+   and timeout.
+1. Improve the Ingress status field to be able to include additional
+   information. The current status currently only contains provisions the IP
+   address of the load balancer.
+1. Formalize the Ingress class annotation into a field and an associated
+   `IngressClass` resource.
+1. Add support for non-Service Backend types.
+
+Note: this is a proposed list. Please discuss in the comments.
+
+### Path as a prefix
+
+The [current APIs][ingress-api] state that the path is a regular expression
+using the [POSIX IEEE Std 1003.1 standard][posix-regex]. However, this is not
+consistent with the syntax supported by any of the common proxy vendors:
+
+| Platform | Syntax                   |
+|----------|--------------------------|
+| nginx    | [PCRE][nginx-re]         |
+| haproxy  | [PCRE/PCRE2][haproxy-re] |
+| envoy    | [ECMAscript][envoy-re]   |
+
+Among cloud providers, there is also inconsistent levels of support
+for regular expression-based path matching. See the load-balancer
+documentation for [aws][aws-re], [gcp][google-re], [azure][azure-re].
+
+It is also the case that our [documentation][ingress-docs] (and most
+Ingress providers) treats the path match as a prefix match. For
+example, a narrow interpretation of the specification would require
+all paths to end with `".*$"`.
+
+#### Proposal
+
+1. Clarify behavior of the existing path regex match and identify that use of
+   path regex is non-portable.
+1. Support prefix matching behavior as the portable alternative.
+
+Change specification of `ingress.spec.rules.http.paths.path` from the current
+text to indicate the non-portability of the value:
+
+> The matching behavior of Path is implementation specific. Path is either a
+> prefix or regular expression match.
+
+Add a field `ingress.spec.rules.http.paths.pathPrefix`:
+
+```golang
+type HTTPIngressPath struct {
+  ...
+  // PathPrefix matches the HTTP request if the request path begins with
+  // this string. For example, PathPrefix = "/foo" will match "/foo", "/foox",
+  // "/foo/bar".
+  //
+  // Both Path and PathPrefix cannot be set (non-empty) at the same time.
+  PathPrefix string
+}
+```
+
+Interoperability between v1beta and v1 will work as follows:
+
+v1beta1 gets the `PathPrefix" field. v1 gets an annotation
```

Comment:
> If we keep both path and pathPrefix in the v1 API then then annotation is not really required, right?
> I'd prefer to push v1 towards "cleaner".

It would be great if we don't use annotations for this.

> Either keep path as an exact match and pathPrefix as a prefix or spec that a trailing "*" differentiates exact-vs-prefix.

The `path` field is being treated as `prefix`, not regex, at least in ingress-nginx, traefik, haproxy, contour, voyager (some of them do not support regular expressions)
In ingress-nginx, we have an [annotation](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#use-regex) to indicate the path should be handled as a regular expression.
I am aware the ingress-gce requires `*` or `.*` at the end of the path.

> First the portable part: Either keep path as an exact match and pathPrefix as a prefix or spec that a trailing "*" differentiates exact-vs-prefix.

What about introducing `pathPrefix` and leave `path` without changes?

Reply:
We need to unambiguosly express an exact match "/foo" or "/foo/" as distinct from a prefix "/foobar" or "/foo/bar".  Whether we do that as 2 fields or a syntax that controllers have to parse is relatively inconsequential in MOST ways.

EXCEPT... there's new, stronger guidance that one-of blocks come with discriminators, so ths would turn into:

```
path: {
  type: Exact
  exact: "/foo"
}
```

and

```
path: {
  type: Prefix
  prefix: "/foo"
}
```

or maybe:

```
{
  pathType: Prefix
  path: "/foo"
}
```
```
{
  pathType: Exact
  path: "/foo"
}
```

This latter leaves room for `type = ImplementationSpecific`, I guess.

The syntax-y way to do this would be something like:

```
{
    path: /foo  # exact
}
```
```
{
    path: /foo*  # prefix
}
```
```
{
    path: impl:/foo  # impl-specific
}
```


Which do we think is most user-friendly?

Reply:
```
{
  pathType: Prefix
  path: "/foo"
}

{
  pathType: Exact
  path: "/foo"
}
```

:+1: from me

> This latter leaves room for type = ImplementationSpecific, I guess.

Also :+1: for that.

Reply:
updating